odra-casper/livenet-env/src/livenet_host_env.rs

```diff
@@ -0,0 +1,174 @@
+    fn set_caller(&self, caller: Address) {
+        self.casper_client.borrow_mut().set_caller(caller);
+    }
+
+    fn set_gas(&self, gas: u64) {
+        self.casper_client.borrow_mut().set_gas(gas);
+    }
+
+    fn caller(&self) -> Address {
+        self.casper_client.borrow().caller()
+    }
+
+    fn get_account(&self, index: usize) -> Address {
+        self.casper_client.borrow().get_account(index)
+    }
+
+    fn balance_of(&self, address: &Address) -> U512 {
+        self.casper_client.borrow().get_balance(address)
+    }
+
+    fn advance_block_time(&self, time_diff: u64) {
+        // Todo: implement logging, especially for such cases:
+        // info!("advance_block_time called - Waiting for {} ms", time_diff);
+        sleep(std::time::Duration::from_millis(time_diff));
+    }
+
+    fn get_event(&self, contract_address: &Address, index: i32) -> Result<Bytes, EventError> {
+        // TODO: handle indices < 0
+        self.casper_client
+            .borrow()
+            .get_event(contract_address, index)
+            .map_err(|_| CouldntExtractEventData)
+    }
+
+    fn get_events_count(&self, contract_address: &Address) -> u32 {
+        self.casper_client.borrow().events_count(contract_address)
+    }
+
+    fn call_contract(
+        &self,
+        address: &Address,
+        call_def: CallDef,
+        use_proxy: bool
+    ) -> Result<Bytes, OdraError> {
+        if !call_def.is_mut() {
+            self.callstack
+                .borrow_mut()
+                .push(CallstackElement::Entrypoint(Entrypoint::new(
+                    *address,
+                    call_def.clone()
+                )));
+            let result = self
+                .contract_register
+                .read()
+                .unwrap()
+                .call(address, call_def);
+            self.callstack.borrow_mut().pop();
+            return result;
+        }
+        match use_proxy {
+            true => self
+                .casper_client
+                .borrow_mut()
+                .deploy_entrypoint_call_with_proxy(*address, call_def),
+            false => {
+                self.casper_client
+                    .borrow_mut()
+                    .deploy_entrypoint_call(*address, call_def);
+                Ok(Default::default())
+            }
+        }
+    }
+
+    fn register_contract(&self, address: Address, entry_points_caller: EntryPointsCaller) {
+        self.contract_register.write().unwrap().add(
+            address,
+            ContractContainer::new(&address.to_string(), entry_points_caller)
+        );
+    }
+
+    fn new_contract(
+        &self,
+        name: &str,
+        init_args: Option<RuntimeArgs>,
+        entry_points_caller: EntryPointsCaller
+    ) -> Address {
+        let mut args = match init_args {
+            None => RuntimeArgs::new(),
+            Some(args) => args
+        };
+        // todo: move this up the stack
+        args.insert("odra_cfg_is_upgradable", false).unwrap();
```

Comment:
Those should be consts from `odra-core`'s consts.

Reply:
Done